File: ios/SproutKit/Tests/SproutDataTests/TreatmentSectionsTest.swift
import XCTest
@testable import SproutData

/// Which side of the treatments list a course lands on — the counterpart of
/// `TreatmentSectionsTest.kt`.
///
/// The end date is inclusive and compared by calendar day, so the boundaries are
/// the whole test: a course whose last dose is today, and one that ended at one
/// minute past midnight.
final class TreatmentSectionsTests: XCTestCase {

    private let calendar = Calendar.current

    /// 15 March 2026, 14:30 local.
    private lazy var today: Date = {
        calendar.date(from: DateComponents(year: 2026, month: 3, day: 15))!
    }()

    private lazy var now: Int64 = millis(at: today, hour: 14, minute: 30)

    private func millis(at day: Date, hour: Int = 12, minute: Int = 0) -> Int64 {
        let point = calendar.date(bySettingHour: hour, minute: minute, second: 0, of: day)!
        return SproutFormat.millis(from: point)
    }

    private func day(offset: Int) -> Date {
        calendar.date(byAdding: .day, value: offset, to: today)!
    }

    private func treatment(endDate: Int64?) -> Treatment {
        Treatment(name: "Vitamin D", startDate: millis(at: day(offset: -30)), endDate: endDate)
    }

    func testNoEndDateIsNeverPast() {
        XCTAssertFalse(hasEnded(treatment(endDate: nil), now: now))
    }

    /// One dose left to give is not something a parent should have to look for
    /// under "finished".
    func testACourseEndingTodayStaysActive() {
        XCTAssertFalse(hasEnded(treatment(endDate: millis(at: today)), now: now))
        XCTAssertFalse(
            hasEnded(treatment(endDate: millis(at: today, hour: 0, minute: 1)), now: now),
            "an instant earlier today is still today"
        )
    }

    func testACourseThatEndedYesterdayIsPast() {
        XCTAssertTrue(hasEnded(treatment(endDate: millis(at: day(offset: -1))), now: now))
        XCTAssertTrue(
            hasEnded(treatment(endDate: millis(at: day(offset: -1), hour: 23, minute: 59)), now: now),
            "right up to the last minute of yesterday"
        )
    }

    func testACourseEndingLaterIsActive() {
        XCTAssertFalse(hasEnded(treatment(endDate: millis(at: day(offset: 1))), now: now))
        XCTAssertFalse(hasEnded(treatment(endDate: millis(at: day(offset: 365))), now: now))
    }

    /// Ahead, not behind. Nothing about a start date can put a course in the
    /// past.
    func testACourseThatHasNotStartedIsActive() {
        let upcoming = Treatment(
            name: "Iron",
            startDate: millis(at: day(offset: 7)),
            endDate: millis(at: day(offset: 21))
        )

        XCTAssertFalse(hasEnded(upcoming, now: now))
    }

    /// The finished ones come out newest first, so the most recently stopped
    /// course is the one at the top of that section.
    func testSectionsSplitAndOrder() {
        let ongoing = treatment(endDate: nil)
        let endsToday = treatment(endDate: millis(at: today))
        let endedLastWeek = treatment(endDate: millis(at: day(offset: -7)))
        let endedYesterday = treatment(endDate: millis(at: day(offset: -1)))

        let sections = treatmentSections(
            [endedLastWeek, ongoing, endedYesterday, endsToday],
            now: now
        )

        XCTAssertEqual(sections.current.map(\.endDate), [ongoing.endDate, endsToday.endDate])
        XCTAssertEqual(
            sections.past.map(\.endDate),
            [endedYesterday.endDate, endedLastWeek.endDate]
        )
    }
}

File: android/app/src/test/java/com/gproust/sprout/TreatmentSectionsTest.kt
package com.gproust.sprout

import com.gproust.sprout.data.local.TreatmentEntity
import com.gproust.sprout.ui.treatments.hasEnded
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId

/**
 * Which side of the treatments list a course lands on. The end date is
 * inclusive and the comparison is by calendar day, so the boundary cases —
 * a course whose last dose is today, one that ends at one minute past
 * midnight — are the ones worth pinning down.
 */
class TreatmentSectionsTest {

    private val zone: ZoneId = ZoneId.systemDefault()
    private val today: LocalDate = LocalDate.of(2026, 3, 15)
    private val now = today.atTime(14, 30).atZone(zone).toInstant().toEpochMilli()

    private fun millisAt(date: LocalDate, time: LocalTime = LocalTime.NOON): Long =
        date.atTime(time).atZone(zone).toInstant().toEpochMilli()

    private fun treatment(endDate: Long?) = TreatmentEntity(
        name = "Vitamin D",
        startDate = millisAt(today.minusDays(30)),
        endDate = endDate,
    )

    @Test
    fun noEndDate_isNeverPast() {
        assertFalse(hasEnded(treatment(endDate = null), now))
    }

    @Test
    fun endsToday_staysActive() {
        assertFalse(hasEnded(treatment(millisAt(today)), now))
        // Even when the stored instant is earlier in the day than "now".
        assertFalse(hasEnded(treatment(millisAt(today, LocalTime.of(0, 1))), now))
    }

    @Test
    fun endedYesterday_isPast() {
        assertTrue(hasEnded(treatment(millisAt(today.minusDays(1))), now))
        // Right up to the last minute of yesterday.
        assertTrue(hasEnded(treatment(millisAt(today.minusDays(1), LocalTime.of(23, 59))), now))
    }

    @Test
    fun endsInTheFuture_isActive() {
        assertFalse(hasEnded(treatment(millisAt(today.plusDays(1))), now))
        assertFalse(hasEnded(treatment(millisAt(today.plusYears(1))), now))
    }

    @Test
    fun notStartedYet_isActiveNotPast() {
        val upcoming = TreatmentEntity(
            name = "Iron",
            startDate = millisAt(today.plusDays(7)),
            endDate = millisAt(today.plusDays(21)),
        )
        assertFalse(hasEnded(upcoming, now))
    }
}
